File: examples/subscriber_example.py
#!/usr/bin/env python3
"""Subscriber example for Zetabus Python library."""

import asyncio
import signal
from src.bus.python.zetabus import ZetabusContext


running = True


def signal_handler(sig, frame):
    global running
    running = False


def message_callback(topic: str, data: bytes):
    """Callback function for received messages."""
    print(f"Received message on topic '{topic}': {data.decode()}")


async def main():
    print("=== Python Subscriber Example ===\n")
    
    # Set up signal handler
    signal.signal(signal.SIGINT, signal_handler)
    
    async with ZetabusContext("nats://localhost:4222") as bus:
        subscriber = await bus.create_subscriber("example.topic", message_callback)
        
        print("Subscribed to topic 'example.topic'")
        print("Waiting for messages... (Press Ctrl+C to exit)\n")
        
        while running:
            await asyncio.sleep(1)
        
        print("\nShutting down...")
        await subscriber.destroy()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/subscriber_example.c
#include "src/bus/c/bus.h"
#include <stdio.h>
#include <unistd.h>
#include <signal.h>

static volatile int running = 1;

void signal_handler(int sig) {
    (void)sig;
    running = 0;
}

void message_callback(const char* topic, const void* data, size_t size) {
    printf("Received message on topic '%s': %.*s\n", topic, (int)size, (const char*)data);
}

int main(void) {
    printf("=== Subscriber Example ===\n\n");

    // Set up signal handler for clean shutdown
    signal(SIGINT, signal_handler);

    zetabus_t* bus = zetabus_create("nats://localhost:4222");
    if (!bus) {
        fprintf(stderr, "Failed to create zetabus\n");
        return 1;
    }

    zetabus_subscriber_t* subscriber = zetabus_subscriber_create(bus, "example.topic", message_callback);
    if (!subscriber) {
        fprintf(stderr, "Failed to create subscriber\n");
        zetabus_destroy(bus);
        return 1;
    }

    printf("Subscribed to topic 'example.topic'\n");
    printf("Waiting for messages... (Press Ctrl+C to exit)\n\n");

    while (running) {
        sleep(1);
    }

    printf("\nShutting down...\n");

    zetabus_subscriber_destroy(subscriber);
    zetabus_destroy(bus);
    return 0;
}
